### PR TITLE
Use relative rpath so builds are reproducible.

### PR DESCRIPTION
**Pull request by jpakkane**
This adds a new argument to rpath evaluation functions, but probably we should get rid of passing build_dir altogether. It's unclear, though, if every build tool can handle relative rpaths (they should, as it's a property of the dynamic linker but you can never be sure about these things).

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -2322,6 +2323,7 @@ def generate_link(self, target, outfile, outname, obj_list, linker, extra_args=[
         return elem
 
     def determine_rpath_dirs(self, target):
+        current_dir = self.get_target_dir(target)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [F841] local variable 'current_dir' is assigned to but never used

**Comment by nirbheek** [comment]
> /usr/bin/ld: warning: libgirlib.so, needed by /root/b mluf9bwy/gir/libgirsubproject.so, not found (try using -rpath or -rpath-link)

Looks like relative paths don't always work?

**Comment by jpakkane** [comment]
This works on Debian Sid. Could this be a case of gir scanner doing poor quoting wrong which is fixed in newer releases? Can someone test this on rawhide for comparison?

**Comment by ignatenkobrain** [comment]
don't have this warning on Fedora Rawhide.

**Comment by jpakkane** [comment]
Also passes on Ubuntu zesty. Is this something rare enough so we can merge this and not worry about it? this only affects building multiple gir files, and those who do that are probably new projects working on up to date platforms.

**Comment by nirbheek** [comment]
I was thinking that if we could figure out what version of the compile-time linker or the runtime linker (or both) added support for this, we could fall back to absolute paths + a warning about reproducible builds. If this didn't work on Zesty, that means it won't work on Xenial, which is an LTS distro and will be very widely used.

**Comment by jpakkane** [comment]
> what version of the compile-time linker or the runtime linker (or both) added support for this

Relative rpaths (using `$ORIGIN`) have worked for 10+ years.

This is most likely not a problem in the linker but in gir scanner. It does not quote `$` so the shell expands it and the rpath used will be wrong. This was fixed at some point in glib AFAICR, but I don't know which version.

> If this didn't work on Zesty

It does work fine on Zesty, but not in Yakkety and therefore not in Xenial either.

**Comment by nirbheek** [comment]
> This was fixed at some point in glib AFAICR, but I don't know which version.

Should look through the logs of [gobject-introspection](https://git.gnome.org/browse/gobject-introspection), then.

> It does work fine on Zesty, but not in Yakkety and therefore not in Xenial either.

Yes, sorry, typo :)

**Comment by jpakkane** [comment]
As a test, could someone run a full GStreamer build with this branch on Xenial just to see if it would work?

**Comment by nirbheek** [comment]
For anyone wanting to help, the way to do this is:

```console
$ git clone git://anongit.freedesktop.org/gstreamer/gst-build
$ cd gst-build/ && meson builddir/ && ninja -C builddir/
```

With that you've built gstreamer. If everything is fine at this point, you will need to also check that the `RPATH` is set correctly and is being used with `ldd` on the gstreamer libs to ensure that the gstreamer libs just built are being used.

**Comment by jpakkane** [comment]
I tested this and ran into problems which begat #1824. Apart from that it seemed to be correct. However between Xenial and Zesty, the toolchain has changed between using rpath and runpath. That might be problematic but has nothing to do with this PR as the behaviour is the same for absolute and relative rpaths.

**Comment by jpakkane** [comment]
Is this something we can merge once #1824 is in? The actual issue seems to affect only few projects and it is only present in old distros.

File: mesonbuild/backend/ninjabackend.py (hunk 1 above)
Comment (nirbheek) [inline mesonbuild/backend/ninjabackend.py]: It seems this flake error was never fixed?

**Comment by nirbheek** [comment]
I'm not sure if I would call Xenial an old distro since it is the current LTS release. Maybe we can turn this on/off depending on the version of glibc in use?

Also on FreeBSD, it seems you also need to pass `-Wl,-z,origin`: https://bugs.documentfoundation.org/show_bug.cgi?id=54015. Maybe that's what you need on Xenial too? What's curious is that `$ORIGIN` has been supported since at least 2008 in glibc. Not sure why it doesn't work on Xenial.

**Comment by jpakkane** [comment]
That freebsd bug is from 2012. It is highly unlikely to be the issue here.

The issue here probably has nothing to do with the core toolchain but some other part (gir scanner?). The only test that fails is `frameworks/12 multiple gir`. That is a relatively recent addition so it probably won't be an issue for projects that are still supporting Xenial et al. It most definitely should not be an issue for GStreamer because if it was it would have caused errors ages ago.

**Comment by tp-m** [comment]
For what it's worth there are certainly some known errors in GStreamer such as #1181 (gir-related).

File: mesonbuild/backend/ninjabackend.py [hunk 2]
```diff
@@ -2307,7 +2307,16 @@ def generate_link(self, target, outfile, outname, obj_list, linker, extra_args=[
         commands += linker.get_option_link_args(self.environment.coredata.compiler_options)
         # Set runtime-paths so we can run executables without needing to set
         # LD_LIBRARY_PATH, etc in the environment. Doesn't work on Windows.
+        if '/' in target.name or '\\' in target.name:
+            # Target names really should not have slashes in them, but
+            # unfortunately we did not check for that and some downstream projects
+            # now have them. Once slashes are forbidden, remove this bit.
+            target_slashname_workaround_dir = os.path.join(os.path.split(target.name)[0],
+                                                       self.get_target_dir(target))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

**Pull request merged**